packages/server/src/fhir/operations/utils/agentutils.ts

```diff
@@ -0,0 +1,202 @@
+    return [badRequest('No agent(s) for given query')];
+  }
+
+  const promises = agents.map((agent: Agent) => handler(agent));
+  const results = await Promise.allSettled(promises);
+  const entries = [] as BundleEntry<Parameters>[];
+
+  for (let i = 0; i < results.length; i++) {
+    const result = results[i];
+    if (result.status === 'rejected') {
+      entries.push(
+        makeResultWrapperEntry(
+          result.reason instanceof OperationOutcomeError ? result.reason.outcome : serverError(result.reason as Error),
+          agents[i]
+        )
+      );
+      continue;
+    }
+    const [outcome, params] = result.value;
+    if (!(isOk(outcome) && params)) {
+      entries.push(makeResultWrapperEntry(outcome, agents[i]));
+      continue;
+    }
+    entries.push(makeResultWrapperEntry(params as Parameters, agents[i]));
+  }
+
+  return [
+    allOk,
+    {
+      resourceType: 'Bundle',
+      type: 'collection',
+      entry: entries,
+    } satisfies Bundle<Parameters>,
+  ];
+}
+
+export function makeResultWrapperEntry(result: Parameters | OperationOutcome, agent: Agent): BundleEntry<Parameters> {
+  return {
+    resource: {
+      resourceType: 'Parameters',
+      parameter: [
+        { name: 'agent', resource: agent },
+        { name: 'result', resource: result },
+      ],
+    },
+  };
+}
+
+export interface AgentMessageOptions {
+  waitForResponse: boolean;
+  timeout?: number;
+}
+
+export async function publishAgentMessage<T extends BaseAgentMessage = BaseAgentMessage>(
+  agent: Agent,
+  message: BaseAgentRequestMessage,
+  options?: AgentMessageOptions
+): Promise<[OperationOutcome] | [OperationOutcome, T | AgentError]> {
+  if (options?.waitForResponse) {
+    let resolve: (response: [OperationOutcome, T | AgentError]) => void;
+    let reject: (err: OperationOutcomeError) => void;
+
+    // If a callback doesn't already exist on the message, tie callback to the associated agent and assign a random ID
+    message.callback ??= getReferenceString(agent) + '-' + randomUUID();
+
+    const redisSubscriber = getRedisSubscriber();
+    await redisSubscriber.subscribe(message.callback);
+
+    redisSubscriber.on('message', (_channel: string, message: string) => {
+      const response = JSON.parse(message) as T | AgentError;
+      resolve([allOk, response]);
+      cleanup();
+    });
+
+    // Create a timer for 5 seconds for timeout
+    const timer = setTimeout(() => {
+      cleanup();
+      reject(new OperationOutcomeError(badRequest('Timeout')));
+    }, options?.timeout ?? 5000);
+
+    const cleanup = (): void => {
+      redisSubscriber.disconnect();
+      clearTimeout(timer);
+    };
+
+    await getRedis().publish(getReferenceString(agent), JSON.stringify(message));
+    return new Promise<[OperationOutcome, T | AgentError]>((_resolve, _reject) => {
+      resolve = _resolve;
+      reject = _reject;
+    });
+  }
+
+  await getRedis().publish(getReferenceString(agent), JSON.stringify(message));
```

Comment:
Recommend we bring back the `publishMessage()` helper which wraps this logic.  The convention of using the agent reference string as the redis channel name might not be totally obvious.